### PR TITLE
[AERIE-1692] Connect Step Simulation to the scheduler

### PR DESCRIPTION
**Pull request by adrienmaillard**
* **Tickets addressed:** AERIE-1692
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
In the first commit, the `IncrementalSimulationDriver` is fixed:
- the exit condition in `simulateSchedule` has been changed: it does not only wait for all tasks to be finished but also all the events that might be immediately/instantly triggered by these activities (changing the value of a register for example). This bug has been found after running a test involving the `banananation` mission model. 
- a private method has been added to simulate until a certain time, without new activities. This method is used to generate `simulationResults` up to a certain time. 

Second commit is making the duration getters return `Optional` so it is cleaner from the outside. 

Third commit is adding a `SimulationFacade` member to `SimResource` to prepare for changes in 5th commit. 

Fourth commit adds the possibility of passing a `SimulationFacade` to the `Problem` ctor as there can be a need for a `SimulationFacade` defined outside the `Problem`, when building a problem P2 from another problem P1 for example as goals defined in P1 are referencing resources from the simulation facade contained in P1.

Fifth commit is the actual connection between step simulation and scheduling : 1) replace `simulatePlan` in `SimulationFacade` by incremental operations `simulateActivity` and `removeActivityFromSimulation` 2) resources are updated only when necessary depending on the queried time domain in addition to the current validity of simulation results making the simulation of empty plans at initialization now useless.

Sixth commit is removing the `SimulationDriverFacade` whose role is handled by `SimulationFacade`. 

Seventh commit is removing the last traces of the merlinsight jar from the gradle file after merge with #63.

## Verification
Scheduler tests, including MerlinsightTests (that require the AerieLander mission model), were run successfully.

## Documentation

Check the step simulation engine for accuracy. 

## Future work

- Caching for resources: when querying SimulationResults, it would be great to be able to query them only for a certain time window. Then, querying the simulation results could also become more incremental.    
- Using a dependency graph to avoid triggering simulation only to get back simulation results

### CODE REVIEW

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (pcrosemurgy) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java:86]: I'd avoid using `assert` since assertion checking can be toggled by the runner of the JVM.
I think it may be best to replace this with: `if (...) throw new Error("...")`.

Also, line 84 seems to have a stray empty comment.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (pcrosemurgy) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: >Second commit is making the duration getters return Optional so it is cleaner from the outside.

Looks like an optional is unnecessary here. Maybe I'm missing some context?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (pcrosemurgy) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: I think the removal of `final` here may have been on accident?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (pcrosemurgy) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Ahh, it looks like you've made lower-level functions return `Optional` to make it easier on higher-level functions. I think any `Optional` wrapping should be done client-side (within the higher-level code), unless `engine.getTaskDuration()` returned a `null` or `Optional` itself.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java:86]: I'll replace the assert with an exception and remove the comment.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Yes it has been an accident, I'll put it again.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (pcrosemurgy) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java:86]: Cool, thanks Adrien!

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: So actually, my code is wrong because the SimulationEngine throws an exception when the duration is not available. I should try/catch it. And what you are saying is that it should be passed up to the simulation facade instead of treated in the driver ?

Comment (pcrosemurgy) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Ohh, good catch, I missed that.

```java
  public Duration getTaskDuration(TaskId taskId){
    final var state = tasks.get(taskId);
    if (state instanceof ExecutionState.Terminated e) {
      return e.joinOffset().minus(e.startOffset());
    }
    throw new IllegalStateException("Asking for the duration of unfinished task");
  }
```

It may be a good idea to change this exception to a custom checked exception (instead of an unchecked runtime exception like it is now). Then you can choose to either bubble up the exception or do a try/catch within `IncrementalSimulationDriver` to wrap it as an `Optional` if the exception is thrown.

All depends on whether the incremental sim. driver is responsible for gracefully handling the exception case or not. I have a feeling that the incremental sim. driver should bubble the exception up and have clients (callers) worry about how it should be handled (`Optional` wrapping, logging an error, etc.).

Comment (mattdailis) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Hmm... I'm not arguing either way, but I do think unconditionally wrapping a value in an Optional is suspicious.

I propose having the simulation engine return an optional.

Just so I understand - under what circumstances would we call getActivityDuration and expect it to be empty?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (mattdailis) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: Just trying to understand - why do we want to allow defining a Problem with no mission model, and no simulation facade?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/SimulationFacade.java
Comment (mattdailis) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/SimulationFacade.java:121]: Just checking - is creating a new IncrementalSimulationDriver the current expected way to reset the simulation?

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: For toy problems, for tests.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/SimulationFacade.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/SimulationFacade.java:121]: The other way was to call `IncrementalSimulation.initSimulation` but it is currently private. Both solutions are very close in terms of what's done.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: @mattdailis this can happen when the simulation of an activity has started but has not finished and you are asking for its duration.

Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Querying the duration of an unfinished task is not a normal operation/a normal interaction to have with the engine, it denotes an algorithmic failure on the client side, that's why I put a runtime exception. What do you think about that ?

Comment (pcrosemurgy) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: So it's not expected that any client (UI or otherwise) will ask for an unfinished task's duration, right? If so then I agree, the `IllegalStateException` runtime ex. sounds okay. But if there's any UI/client-side workflow that regularly queries for an unfinished task's duration then a checked exception would probably be best.

Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: It is not expected no. I'll let the exception in the engine and remove the Optional in the client.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (mattdailis) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: I would like to find a way to make the interface consistent for "real" problems and "test" problems. Not sure off the top of my head how to do that.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Actually... querying a duration of an unfinished task is kind of the only way of knowing whether a task has finished or not. The other way is computing `SimulationResults` but this is very costly. 
I retract my previous statement and I am going with Matt's proposition, the engine returning an `Optional` directly.  
Sorry for the back and forth.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java
Comment (mattdailis) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/Problem.java]: I made a ticket for this, no need to do this in this PR: AERIE-1748

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (mattdailis) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Looks good to me

**Review by mattdailis — APPROVED** [review]
I have no remaining concerns for this PR

**Comment by mattdailis** [comment]
The plan is to merge #54 before this PR, so that @adrienmaillard can rebase onto it.

**Comment by pcrosemurgy** [comment]
@mattdailis looks like #54 is marked as "do not merge"

**Review by pcrosemurgy — APPROVED** [review]
Thanks Adrien!

**Comment by mattdailis** [comment]
> @mattdailis looks like #54 is marked as "do not merge"

Yes - you're right. I marked it with that because although it has two approvals, I haven't tested a key piece of its functionality yet. In particular, I want to make sure I got the JAR loading logic correct for the new `SchedulerPlugin`. I'm working on testing that right now, and if it works I'll remove the label and merge that PR :+1:

**Comment by pcrosemurgy** [comment]
Ahh got it, good thinking!

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (kroffo) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: This variable is computed up here, but not used until line 95, the only place it is used. I suggest you either move this to just above line 94, or swap lines 94 and 95 and inline this computation.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (kroffo) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: Just a naming suggestion, you don't have to go with this if you don't agree, but I think a more fitting name for this method may be `getSimulationResultsUpTo`.

**Review by kroffo — APPROVED** [review]
Some minor comments, but overall looks good. Because the issues I commented are so minor, I've gone ahead and given approval, but please do look at them and consider fixing if you agree with them.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java
Comment (adrienmaillard) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java]: I like it! I'll rename this method before merging.

**Pull request merged**
